src/process.c

```diff
@@ -778,6 +778,246 @@ cc_free_pointer(gpointer str)
 	g_free(str);
 }
 
+/*
+ * Hash to map a MAC address to a given BDF.  This is populated and used
+ * to identify Virtual Function networking devices which are meant to be
+ * passed to QEMU
+ */
+GHashTable* mac_hash;
+
+/*!
+ * Function to store a BDF value for a given MAC address
+ *
+ * \param hash \ref GHashTable
+ * \param mac \ref gpointer
+ * \param bdf \ ref gpointer
+ */
+static void
+hash_mac_and_store_bdf(GHashTable* hash, gpointer mac, gpointer bdf)
+{
+       g_hash_table_insert(hash, mac, bdf);
+}
+
+/*!
+ * Helper function to get a MAC and BDF for a given network device.
+ * While parsing the sysfs, also grabbing vendor/device ID and driver
+ * associated with the device
+ *
+ * \param dirname \ref gchar*
+ * \param dev_info \ref cc_oci_device to return BDF and driver info
+ * \param mac \ ref gchar** mac address to return found MAC into
+ *
+ * \return \c true if succesfully found mac/bdf
+ */
+static gboolean
+get_mac_and_bdf(gchar *dirname, struct cc_oci_device *dev_info, gchar **mac)
+{
+	gchar *tmp_path = NULL;
+	gchar *device_dir_path = NULL;
+	gchar *buffer = NULL;
+	gchar *sym_link = NULL;
+	GError *link_error = NULL;
+	gchar **tokens = NULL;
+	gchar **driver_tokens = NULL;
+	gchar **mac_from_file = NULL;
+	gchar **device_id_from_file = NULL;
+	gchar **vendor_id_from_file = NULL;
+	gchar *vendor_device_id = NULL;
+
+	gboolean retval = false;
+	guint idx=0;
+
+	/* grab the sym_link for the given device.  dirname is the
+	 * particular device name @ /sys/class/net/
+	 */
+	tmp_path = g_strdup_printf("/sys/class/net/%s", dirname);
+
+	sym_link = g_file_read_link(tmp_path, &link_error);
+	if (!sym_link) {
+		if (link_error) {
+			g_debug ("sym_link read failure. Path: %s", tmp_path);
+		}
+		goto out;
+	}
+
+	/*
+	 * parse the sym_link in order to grab the bdf and make sure it is a
+	 * pci device.  If it is a virtual device, skip out.
+	 */
+	tokens = g_strsplit (sym_link, "/", 8);
+	if (!tokens) {
+		g_debug("sym_link read failure!");
+		goto out;
+	}
+
+	if ( !(g_strcmp0(tokens[3], "virtual")) ) {
+		g_debug("virtual device found");
+		goto out;
+	}
+
+	/* Grab BDF */
+	device_dir_path = g_strdup_printf("/sys/class/net/%s/device", sym_link);
+	sym_link = g_file_read_link(device_dir_path, &link_error);
+	if (!sym_link) {
+		if (link_error) {
+			g_debug ("device relative-path sym_link read failure. Path: %s", tmp_path);
+		}
+		goto out;
+	}
+
+	tokens = g_strsplit (sym_link, "/", 4);
+	if (!tokens || !tokens[3] ) {
+		g_debug("unexpected device path format\n");
+		goto out;
+	}
+
+	/* Grab device ID */
+	tmp_path = g_strdup_printf("%s/device", device_dir_path);
+	if ( !(g_file_get_contents(tmp_path, &buffer, NULL, NULL)) ) {
+		g_debug("Reading file %s returned error", tmp_path);
+		goto out;
+	}
+	g_free_if_set(tmp_path);
```

Comment:
There are a lot of calls to `g_free()` and `g_free_if_set()` in this function which makes it look quite "busy". You might want to declare these variables like this to avoid having to constantly free their values:

```
g_autofree gchar  *foo = NULL;
```